client/conn.go

```diff
@@ -489,12 +495,62 @@ func (c *Conn) ReadOKPacket() (*Result, error) {
 	return c.readOK()
 }
 
+// Send COM_QUERY and read the result
 func (c *Conn) exec(query string) (*Result, error) {
-	if err := c.writeCommandStr(COM_QUERY, query); err != nil {
+	err := c.execSend(query)
+	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	return errors.Trace(c.readResult(false))
+}
+
+// Sends COM_QUERY
+// https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_com_query.html
+func (c *Conn) execSend(query string) error {
+	var buf bytes.Buffer
+
+	if c.includeLine >= 0 {
+		_, file, line, ok := runtime.Caller(c.includeLine)
+		if ok {
+			lineAttr := QueryAttribute{
+				Name:  "_line",
+				Value: fmt.Sprintf("%s:%d", file, line),
+			}
+			c.queryAttributes = append(c.queryAttributes, lineAttr)
+		}
+	}
+
+	if c.capability&CLIENT_QUERY_ATTRIBUTES > 0 {
+		numParams := len(c.queryAttributes)
+		buf.Write(PutLengthEncodedInt(uint64(numParams)))
+		buf.WriteByte(0x1) // parameter_set_count, unused
+		if numParams > 0 {
+			// null_bitmap, length: (num_params+7)/8
+			for i := 0; i < (numParams+7)/8; i++ {
+				buf.WriteByte(0x0)
+			}
+			buf.WriteByte(0x1) // new_params_bind_flag, unused
+			for _, qa := range c.queryAttributes {
+				buf.Write(qa.TypeAndFlag())
+				buf.Write(PutLengthEncodedString([]byte(qa.Name)))
+			}
+			for _, qa := range c.queryAttributes {
+				buf.Write(qa.ValueBytes())
+			}
+		}
+	}
+
+	_, err := buf.Write(utils.StringToByteSlice(query))
+	if err != nil {
+		return err
+	}
+
+	if err := c.writeCommandBuf(COM_QUERY, buf.Bytes()); err != nil {
+		return errors.Trace(err)
+	}
+	c.queryAttributes = nil
```

Comment:
I prefer we use `clear(c.queryAttributes)` to reuse the slice to reduce the GC pressure.

And use `defer` to clear it in any cases